### PR TITLE
Add keybind to checkout commit in log view

### LINKED ISSUE #983: [Feature] Allow checking out specific commits in log view
Hi! I just found this project and think it's pretty cool!

Anyway, I think it would be nice if the log view had an option to enter detached HEAD state and checkout the selected commit.

### PR DESCRIPTION
**Pull request by Omnikar**
This Pull Request fixes/closes #983.

It changes the following:
- Adds `Shift-S` keybind to checkout the selected commit in the log view.

I followed the checklist:
- [ ] I added unittests
- [x] I ran `make check` without errors
- [x] I tested the overall application
- [x] I added an appropriate item to the changelog

I think it'd be better if I could add a confirmation before detaching HEAD, similar to the one used for branch deletion and such, but I'm not sure how to.

### CODE REVIEW

File: src/components/commitlist.rs [hunk 1]
```diff
@@ -423,29 +426,40 @@ impl DrawableComponent for CommitList {
...
-				};
+					checkout_commit(
+						&self.repo.borrow(),
+						commit_hash,
+					)?;
+				}
+				true
+			} else {
+				false
+			};
```
Comment (Omnikar) [inline src/components/commitlist.rs:462]: Rustfmt decided to behave rather strangely here.

Comment (extrawurst) [reply src/components/commitlist.rs:462]: lgtm

File: src/components/commitlist.rs [hunk 2]
```diff
@@ -423,29 +426,40 @@ impl DrawableComponent for CommitList {
...
+			} else if k == self.key_config.keys.page_up {
+				self.move_selection(ScrollType::PageUp)?
+			} else if k == self.key_config.keys.page_down {
+				self.move_selection(ScrollType::PageDown)?
+			} else if k == self.key_config.keys.log_mark_commit {
+				self.mark();
+				true
+			} else if k == self.key_config.keys.log_checkout_commit {
+				if let Some(commit_hash) =
+					self.selected_entry().map(|entry| entry.id)
```
Comment (extrawurst) [inline src/components/commitlist.rs:452]: lets put this into a `self.checkout`

File: asyncgit/src/sync/branch/mod.rs [hunk 3]
```diff
@@ -301,6 +301,36 @@ pub fn checkout_branch(
...
+	let repo = repo(repo_path)?;
+	let cur_ref = repo.head()?;
+	let statuses = repo.statuses(Some(
+		git2::StatusOptions::new().include_ignored(false),
+	))?;
+
+	if statuses.is_empty() {
+		repo.set_head_detached(commit_hash.into())?;
+
+		if let Err(e) = repo.checkout_head(Some(
```
Comment (extrawurst) [inline asyncgit/src/sync/branch/mod.rs:320]: there is a bunch of code duplication with `checkout_remote_branch` and `checkout_branch` - can reuse something?

File: asyncgit/src/sync/branch/mod.rs [hunk 4]
```diff
@@ -301,6 +301,36 @@ pub fn checkout_branch(
 	}
 }
 
+/// Detach HEAD to point to a commit then checkout HEAD, does not work if there are uncommitted changes
+pub fn checkout_commit(
```
Comment (extrawurst) [inline asyncgit/src/sync/branch/mod.rs:305]: we need unittest(s) for `checkout_commit`

**Review by extrawurst — CHANGES_REQUESTED** [review]

File: asyncgit/src/sync/branch/mod.rs (hunk 4 above)
Comment (Omnikar) [reply asyncgit/src/sync/branch/mod.rs:305]: Where should I put it?

**Comment by Omnikar** [comment]
How do I get the keybind to appear in this bar at the bottom?
image

File: asyncgit/src/sync/branch/mod.rs (hunk 4 above)
Comment (extrawurst) [reply asyncgit/src/sync/branch/mod.rs:305]: check the bottom of the file, there should be a few examples of tests for the other functions. use `cargo t --workspace` to run all tests and once ready run `make check` to validate what will be run on the CI

**Comment by extrawurst** [comment]
> How do I get the keybind to appear in this bar at the bottom?

checkout the `commands` fn next to `events` to see how components add their commands

**Comment by extrawurst** [comment]
@Omnikar feel free to reach out on our discord aswell

**Comment by extrawurst** [comment]
@Omnikar are you planning to follow this up?

File: asyncgit/src/sync/branch/mod.rs (hunk 3 above)
Comment (Omnikar) [reply asyncgit/src/sync/branch/mod.rs:320]: I'm not sure exactly what I could extract, what do you suggest?

File: asyncgit/src/sync/branch/mod.rs (hunk 4 above)
Comment (extrawurst) [reply asyncgit/src/sync/branch/mod.rs:305]: Ping?

**Comment by fralcow** [comment]
Hey @Omnikar, I'm willing to pick this issue up. I see you left a couple of comments to the original author regarding unit tests and code reuse, should I pick up their branch and fix this stuff, or would you suggest writing the whole thing from scratch?

**Comment by fralcow** [comment]
Hey @extrawurst , I'm willing to pick this issue up. I see you left a couple of comments to the original author regarding unit tests and code reuse, should I pick up their branch and fix this stuff, or would you suggest rewriting the whole thing from scratch?

**Pull request closed without merging**